### DISCUSSION TITLE
Discrepancy between Node.js version in .nvmrc and `engines` field in package.json after Renovate updates

### DISCUSSION THREAD

**Discussion by liby** [renovatebot/renovate#22552] [Request Help] [answered]
### How are you running Renovate?

Placeholder value, please select the correct response from the dropdown

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

Placeholder value, please select the correct response from the dropdown

### Describe the problem

I've recently encountered an issue where Renovate has updated the Node.js versions in my _.nvmrc_ and _package.json_  files, and now they are significantly different.

Specifically, the Node.js version in my _.nvmrc_ file was initially updated to `16.6.0`, while the `engines` field in my _package.json_ was updated to `20.2.0` by Renovate.

Later, Renovate opened another PR to update the Node.js version in _.nvmrc_ to the LTS version, but even after that, there is still a discrepancy between the versions in _.nvmrc_ and _package.json_.

I would like to maintain the Node.js version at its LTS version in both files. However, Renovate seems to be updating the version beyond the LTS version, which has caused this discrepancy.

Is there a way to configure Renovate or my project setup to ensure that the Node.js versions in these two files are aligned, and that they are updated only up to the latest LTS version?

### Relevant debug logs

<details><summary>Related Log sections</summary>

```
DEBUG: packageFiles with updates (repository=7, baseBranch=develop)
       "config": {
         "gitlabci": [
           {
             "deps": [
               ...
               {
                 "commitMessageTopic": "Node.js",
                 "currentValue": ">=16.0.0",
                 "currentVersion": "v20.2.0",
                 "datasource": "github-tags",
                 "depName": "node",
                 "depType": "engines",
                 "lockedVersion": null,
                 "packageName": "nodejs/node",
                 "prettyDepType": "engine",
                 "registryUrl": "https://github.com",
                 "sourceUrl": "https://github.com/nodejs/node",
                 "versioning": "node",
                 "warnings": [],
                 "updates": [
                   {
                     "updateType": "pin",
                     "isPin": true,
                     "newValue": "20.2.0",
                     "newVersion": "v20.2.0",
                     "newMajor": 20,
                     "branchName": "feature/renovate-all-minor-patch-pin"
                   }
                 ]
               },
               {
                 "commitMessageTopic": "pnpm",
                 "currentValue": "8.5.1",
                 "currentVersion": "8.5.1",
                 "datasource": "npm",
                 "depName": "pnpm",
                 "depType": "packageManager",
                 "fixedVersion": "8.5.1",
                 "homepage": "https://pnpm.io",
                 "isSingleVersion": true,
                 "lockedVersion": null,
                 "packageName": "pnpm",
                 "prettyDepType": "packageManager",
                 "registryUrl": "https://registry.npmjs.org",
                 "sourceUrl": "https://github.com/pnpm/pnpm",
                 "versioning": "npm",
                 "warnings": [],
                 "updates": [
                   {
                     "bucket": "non-major",
                     "newVersion": "8.6.0",
                     "newValue": "8.6.0",
                     "releaseTimestamp": "2023-05-28T12:18:41.032Z",
                     "newMajor": 8,
                     "newMinor": 6,
                     "updateType": "minor",
                     "branchName": "feature/renovate-all-minor-patch-pin"
                   }
                 ]
               },
               {
                 "currentValue": "fonts/*",
                 "depName": "main",
                 "depType": "overrides",
                 "lockedVersion": null,
                 "managerData": {"parents": ["@bower_components/bootstrap"]},
                 "packageName": "main",
                 "skipReason": "unspecified-version",
                 "updates": []
               }
             ],
             "lockFiles": ["pnpm-lock.yaml"],
             "managerData": {
               "hasPackageManager": true,
               "packageJsonName": "company-name-frontend",
               "pnpmShrinkwrap": "pnpm-lock.yaml",
               "yarnZeroInstall": false
             },
             "npmrc": "***********",
             "packageFile": "package.json",
             "packageFileVersion": "1.0.0",
             "skipInstalls": true
           }
         ],
         "nvm": [
           {
             "deps": [
               {
                 "currentValue": "16",
                 "currentVersion": "v16.20.0",
                 "datasource": "github-tags",
                 "depName": "node",
                 "isSingleVersion": false,
                 "packageName": "nodejs/node",
                 "registryUrl": "https://github.com",
                 "sourceUrl": "https://github.com/nodejs/node",
                 "versioning": "node",
                 "warnings": [],
                 "updates": [
                   {
                     "updateType": "pin",
                     "isPin": true,
                     "newValue": "16.20.0",
                     "newVersion": "v16.20.0",
                     "newMajor": 16,
                     "branchName": "feature/renovate-all-minor-patch-pin"
                   },
                   {
                     "bucket": "major",
                     "newVersion": "v18.16.0",
                     "newValue": "18",
                     "releaseTimestamp": "2023-04-12T23:16:25.000Z",
                     "newMajor": 18,
                     "newMinor": 16,
                     "updateType": "major",
                     "isRange": true,
                     "branchName": "feature/renovate-node-18.x"
                   }
                 ]
               }
             ],
             "packageFile": ".nvmrc"
           }
         ]
       }
```

</details>

[full.log](https://github.com/renovatebot/renovate/files/11632975/full.log)

### Have you created a minimal reproduction repository?

N/A

**Comment by rarkins** [comment, marked as answer]
According to the log, your package.json engines has node as `">=16.0.0"` which means it's already allowing v20. All Renovate is doing is proposing _pinning_ it to an exact 20.x version (highest satisfying). Meanwhile `.nvmrc` has value `"16"` so Renovate proposes pinning it to an exact 16.x version.

If you have turned on version pinning, it sounds like you should turn it off (partly because you don't seem to understand it, and it's best to leave features off unless you understand them). The engines field in package.json should never be pinned.

**Reply by liby** [reply]
Thank you for your response. 

We were considering a solution which involves adding a package rule specific to Node.js in the Renovate config. The rule would look something like this:

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": ["config:base"],
  "rangeStrategy": "pin",
  "platformAutomerge": false,
  "prCreation": "immediate",
  "prHourlyLimit": 1,
  "packageRules": [
    {
      "groupName": "Dependencies (non-major)",
      "groupSlug": "all-minor-patch-pin",
      "matchUpdateTypes": ["patch", "minor", "pin"]
    },
    {
      "description": "Node.js version constraint",
      "matchPackageNames": ["node"],
      "rangeStrategy": "replace",
      "allowedVersions": ">=16"
    }
  ]
}
```

In this rule, we aim to replace the Node.js version with a version that's equal or higher than 16, instead of pinning it to a specific version. Is this solution feasible?

Thank you in advance for your insight.

**Reply by rarkins** [reply]
`rangeStrategy=pin` at the top level of your config is likely a bad idea, for example you rarely ever would want to pin engines in package.json.

I also recommend against grouping pin PRs with patch or minor

**Reply by liby** [reply]
Thank you for your response and suggestions. I agree with your points and will change the `rangeStrategy` to "bump" at the top level of the configuration.

As for grouping "pin" with "patch" and "minor", the primary reason for this configuration was to handle the case where any 0.y.z or 0.0.z updates are considered major updates, even if the y or z value changes. In our context, we don't regard such updates as major, and we prefer not to open a new Merge Request for every patch or minor upgrade for each package.

From your advice, it appears that this approach might not be suitable. Would you recommend a better practice for handling this specific scenario? We'd really appreciate any guidance you could provide.

**Reply by rarkins** [reply]
Careful not to bump peer dependencies or engines fields - you probably don't want that. Pinning is otherwise fine and in fact I recommend it for the majority of dependencies.

Pin PRs are not "updates" - their aim is just to narrow any existing range to a single version. They are already grouped together by default so you should see one big PR at the start and then have it rarely happen after that (only in case you add a dependency to the project with a range).

Regarding 0.x updates, here is an example of a package rule which excludes them from other patch and minor: https://docs.renovatebot.com/key-concepts/automerge/#automerge-non-major-updates

**Reply by liby** [reply]
I have a question regarding the handling of pre-1.0.0 updates in Renovate. According to the SemVer specification, pre-1.0.0 versions can make breaking changes at any time. Therefore, I was under the impression that "patch" and "minor" updates would not match these pre-1.0.0 updates, and that they would always be considered as "major" updates by Renovate, resulting in a separate Merge Request.

However, based on a recent discussion, it seems that a package rule with "patch" and "minor" in `matchUpdateTypes` would indeed match updates like `0.0.1` to `0.0.2` or `0.0.x` to `0.1.0`. Could you please confirm if my understanding is correct?

Here is my updated configuration:

```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": ["config:base", ":pinAllExceptPeerDependencies"],
  "platformAutomerge": false,
  "prCreation": "immediate",
  "prHourlyLimit": 1,
  "packageRules": [
    {
      "groupName": "Dependencies (non-major)",
      "groupSlug": "all-minor-patch-pin",
      "matchUpdateTypes": ["patch", "minor"]
    }
  ]
}
```

In the new configuration, I removed `rangeStrategy` from the top level and instead extended [`:pinAllExceptPeerDependencies](https://docs.renovatebot.com/presets-default/#pinallexceptpeerdependencies)`. This will pin all dependencies while excluding `engines` and `peerDependencies`.

I also removed "pin" from `matchUpdateTypes`. Based on my understanding of our previous discussion, "patch" and "minor" can match updates prior to pre-1.0.0, and Renovate will not open a separate Merge Request for these updates.

**Reply by rarkins** [reply]
That question is getting too off-topic for this original thread. Please start a new Discussion
